memory_w25q1128jv_spi/memory_w25q1128jv_spi.cc

```diff
@@ -0,0 +1,220 @@
+    if (DeviceID() != 0xEF17) return -1;
+
+    // do while to find the first free 64 byte address
+    do {
+        address = ChipReadDump(buffer);
+    } while (buffer[0] != 0xFF);
+
+    return 0;
+}
+
+int MemoryW25q1128jvSpi::DeviceID() {
+    // Control pins his for no false memory actions
+    HAL_GPIO_WritePin(_wpPort, _wpPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_holdPort, _holdPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // input instruction hex + 3 dummy bytes
+    uint8_t DEVICE_ID[4] = {0x90, 0x00, 0x00, 0x00};
+    // output array of bytes
+    uint8_t deviceValue[2];
+
+    // SPI protocol
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(_hspi, DEVICE_ID, 4, SPI_Timeout) != HAL_OK) return -1;
+    if (HAL_SPI_Receive(_hspi, deviceValue, 2, SPI_Timeout) != HAL_OK) return -1;
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // returns the 2 bytes as a uint16_t
+    return (int)(deviceValue[0] << 8 | deviceValue[1]);
+}
+
+int MemoryW25q1128jvSpi::ReadStatusReg1() {
+    // Control pins his for no false memory actions
+    HAL_GPIO_WritePin(_wpPort, _wpPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_holdPort, _holdPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // input operation code for reading status register 1
+    uint8_t READ_STATUS_REG_1[1] = {0x05};
+    // output -- 8 bit value of the status register
+    uint8_t regValue[1];
+
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(_hspi, READ_STATUS_REG_1, 1, SPI_Timeout) != HAL_OK) return -1;
+    if (HAL_SPI_Receive(_hspi, regValue, 1, SPI_Timeout) != HAL_OK) return -1;
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    return (int)regValue[0];
+}
+
+int MemoryW25q1128jvSpi::ChipErase() {
+    // Control pins his for no false memory actions
+    HAL_GPIO_WritePin(_wpPort, _wpPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_holdPort, _holdPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    /* Set Write Enable to 1 */
+    uint8_t CHIP_ENABLE[1] = {0x06};
+
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(_hspi, CHIP_ENABLE, 1, SPI_Timeout) != HAL_OK) return -1;
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+    ;
+    while ((MemoryW25q1128jvSpi::ReadStatusReg1() & 0x02) != 2) {
+    }
+
+    /* Initiate Chip Erase */
+    uint8_t CHIP_ERASE[1] = {0xC7};
+
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(_hspi, CHIP_ERASE, 1, SPI_Timeout) != HAL_OK) return -1;
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // Waits until busy bit is = 0
+    // write busy is 1st bit in status reg
+    while ((MemoryW25q1128jvSpi::ReadStatusReg1() & 0x01) != 0) {
+    }
+
+    return 0;
+}
+
+MemoryW25q1128jvSpi::State MemoryW25q1128jvSpi::ChipWrite(uint8_t (&data)[64]) {
```

Comment:
this `uint8_t (&data)[64]` is uhhhh, unusual? i've never seen this before, `uint8_t *data` is fine.
is this some fancy c++ thing im not aware of?

Reply:
Yea, I looked up this one as a solution. `uint8_t *data` is way more standard to see, but I just wanted to specify that the array to be sent into the function has to be exactly 64 elements. We can probably change the code to `uint8_t *data` if you don't want to the end user to be errored out by the compiler when they try to send an array of different size than 64 byes, but personally, I like this implementation.